airflow/contrib/operators/neo4j_operator.py

```diff
@@ -0,0 +1,107 @@
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Neo4JOperator to interact and perform action on Neo4J graph database.
+This operator is designed to use Neo4J Hook and the
+Python driver: https://neo4j.com/docs/api/python-driver/current/
+"""
+
+import csv
+
+from neo4j import BoltStatementResult
+
+from airflow.contrib.hooks.neo4j_hook import Neo4JHook
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+
+
+class Neo4JOperator(BaseOperator):
+    """
+    This operator takes a number of configuration parameters
+    :param cypher_query: required cypher query to be executed on the Neo4J database
+    :type cypher_query: str
+    :param output_filename: required filename to produce with output from the query
+    :type output_filename: str
+    :param n4j_conn_id: reference to a pre-defined Neo4J Connection
+    :type n4j_conn_id: str
+    :param fail_on_no_results: True/False flag to indicate if it should fail the task if no results
+    :type fail_on_no_results: bool
+    """
+    _cypher_query = None
+    _output_filename = None
+    _n4j_conn_id = None
+    _fail_on_no_results = None
+
+    template_fields = ['cypher_query', 'output_filename', 'n4j_conn_id']
+
+    @apply_defaults
+    def __init__(self,
+                 cypher_query,
+                 output_filename,
+                 n4j_conn_id,
+                 fail_on_no_results=False,
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._output_filename = output_filename
+        self._cypher_query = cypher_query
+        self._n4j_conn_id = n4j_conn_id
+        self._fail_on_no_results = fail_on_no_results
+
+    def execute(self, context):
+        """
+        Executes the supplied query and saves the results as a CSV file on disk
+        :param context:
+        :return:
+        """
+        hook = Neo4JHook(n4j_conn_id=self._n4j_conn_id)
+        result: BoltStatementResult = hook.run_query(cypher_query=self._cypher_query)
+
+        # In some cases, an empty result should fail (where results are expected)
+        if result.peek() is None and self._fail_on_no_results:
+            raise AirflowException("Query returned no rows")
+
+        row_count = self._make_csv(result)
+
+        # Provide some feedback to what was done...
+        self.log.info("Saved %s with %s rows", self._output_filename, row_count)
+
+    def _make_csv(self, result: BoltStatementResult):
```

Comment:
Shouldn't this method be in the hook?

Reply:
Excellent point. I think of the hook as providing the descrete mechanisim to connect to the database. It is a wrapper for the library that does this.
There are a few other operators that handle file writing operations like vertica_to_mysql and sql_to_gcs and so on. Happy to move it, but the hook didnt feel like the right place for it either.

Reply:
Think of case where you would want to export neo4j data (for backup or other)
now you can export it to local file, GCS, S3, FTP and the list is long.
If this function remains on the Operator each one will need to duplicate the logic.
On the other hand if the function is on the hook all operators that use the hook will be able to use it.

This is why most of the db hook has this method in them (and other like get_pandas_df )

By the way most hooks call this method `to_csv` better to stay consistent.

Reply:
Thankyou @OmerJog 
I have moved this from the operator to the hook as you suggested. I could not find any other examples of this, but happy to take your advice on board.